dbconn/reload_credentials.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2022 Block, Inc.
+
+package dbconn
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	dsndriver "github.com/go-mysql/hotswap-dsn-driver"
+	"github.com/go-sql-driver/mysql"
+
+	"github.com/cashapp/blip"
+	"github.com/cashapp/blip/event"
+)
+
+func init() {
+	dsndriver.SetHotswapFunc(Repo.ReloadDSN)
+}
+
+type Credentials struct {
+	Hostname string
+	Username string
+	Password string
+	TLS      *blip.ConfigTLS
+}
+
+type CredentialFunc func(context.Context) (*Credentials, error)
+
+type repo struct {
+	m       *sync.Map
+	tlsHash *sync.Map
+}
+
+var Repo = &repo{
+	m:       &sync.Map{},
+	tlsHash: &sync.Map{},
+}
+
+func (r *repo) Add(addr string, f CredentialFunc, tlsHash string) error {
```

Comment:
Why is `tlsHash` not a field in `Credentials`? Seems like it's just another part of the credentials.

Reply:
I could add the hash as a field on `Credentials` so it's computed when pulling the password info, if needed. For the `Add` method I'm leaning away from invoking `CredentialsFunc` to get the hash though, which would mean I'd leave the function signature as-is. Do you have a preference here?

Reply:
Alternatively, if we want the hash to be computed when loading the certs then this makes less sense as `CredentialsFunc` is just parsing the CNF file, not loading config based on the CNF file contents. To me feels like it makes more sense to have `CredentialsFunc` load the CNF file contents, which we then can check for changes in `ReloadDSN`. If we detect changes we load the TLS files. Granted we do have to parse the TLS files to generate the hash, but we could simply do that and load the TLS settings from the files at the same time and discard them if there are no changes.